blockbuilder/blockbuilder.go

```diff
@@ -0,0 +1,249 @@
+
+const (
+	BlockGasTargetDivisor uint64 = 1024 // The bound divisor of the gas limit, used in update calculations
+)
+
+// BlockBuilderParams are fields for the block that cannot be changed
+type BlockBuilderParams struct {
+	// Parent block
+	Parent *types.Header
+
+	// Executor. TODO: Abstract executor
+	Executor *state.Executor
+
+	// Coinbase that is signing the block
+	Coinbase types.Address
+
+	// Vanity extra for the block
+	VanityExtra []byte
+
+	// GasLimit is the gas limit for the block
+	BlockGasTarget uint64
+
+	// Logger
+	Logger hclog.Logger
+
+	// txPoolInterface implementation
+	TxPool txPoolInterface
+}
+
+type blockchainInterface interface {
+	// WriteBlock writes the block to the db but does not process it
+	WriteBlock(b *types.Block, source string) error
+}
+
+type txPoolInterface interface {
+	Prepare()
+	Peek() *types.Transaction
+	Pop(tx *types.Transaction)
+	Drop(tx *types.Transaction)
+	Demote(tx *types.Transaction)
+}
+
+func NewBlockBuilder(params *BlockBuilderParams) (*BlockBuilder, error) {
+	// vanity extra can only be 32 size max. it is better to trim that to return
+	// an error that we have to propagate. It should be up to higher level
+	// code to error if the vanity extra supplied by the user is too big.
+	if len(params.VanityExtra) > 32 {
+		params.VanityExtra = params.VanityExtra[:32]
+	}
+
+	if params.VanityExtra == nil {
+		params.VanityExtra = make([]byte, 0)
+	}
+
+	if params.Logger == nil {
+		params.Logger = hclog.NewNullLogger()
+	} else {
+		params.Logger = params.Logger.Named("block-builder")
+	}
+
+	builder := &BlockBuilder{
+		params: params,
+	}
+	if err := builder.Reset(); err != nil {
+		return nil, err
+	}
+
+	return builder, nil
+}
+
+type BlockBuilder struct {
+	// input params for the block
+	params *BlockBuilderParams
+
+	// header is the header for the block being created
+	header *types.Header
+
+	// transactions are the data included in the block
+	txns []*types.Transaction
+
+	// block is a reference to the already built block
+	block *types.Block
+
+	state *state.Transition
+}
+
+func (b *BlockBuilder) Reset() error {
+	// Generate the base block
+	header := &types.Header{
+		ParentHash: b.params.Parent.Hash,
+		Number:     b.params.Parent.Number + 1,
+		Miner:      types.ZeroAddress.Bytes(),
+		Nonce:      types.Nonce{},
+		MixHash:    signer.IstanbulDigest,
+		Difficulty: b.params.Parent.Number + 1,
+		StateRoot:  types.EmptyRootHash,
+		Sha3Uncles: types.EmptyUncleHash,
+		GasLimit:   b.params.Parent.GasLimit,
+	}
+
+	b.block = nil
+	b.txns = []*types.Transaction{}
+
+	// calculate gas limit based on parent header
+	header.GasLimit = calculateGasLimit(b.params.Parent.GasLimit, b.params.BlockGasTarget)
+
+	transition, err := b.params.Executor.BeginTxn(b.params.Parent.StateRoot, b.header, b.params.Coinbase)
+	if err != nil {
+		return err
+	}
+
+	b.state = transition
+
+	return nil
+}
+
+// Fill fills the block with transactions from the txpool
+func (b *BlockBuilder) Fill() error {
```

Comment:
Basically, `NewBlockBuilder()`, `Fill()`, and `BuildBlock()` seems to be called sequentially. How about defining helper function that takes argument and reference and returns constructed block?

Reply:
in old v3, `Fill` was called inside `BuildBlock`, i think. but I think we should be able to call `NewBlockBuilder` and `BuildBlock` separately (at least for polybft because we are not creating and building from the same place)

Reply:
`NewBlockBuilder` does not need to be instantiated for each built block. It is done as part of this PR for simplicity. `v3-parity` branch has only one instance per sequence that gets reset.

As for `Fill` and `Build`. In `v3-parity` there is an extra step in between to include custom transactions (state transactions) before the block gets built. So, it does not make sense to unify both.